### ISSUE TITLE
MP4 video not playing

### ISSUE THREAD

**Issue by oliversturm** [qutebrowser/qutebrowser#4085]
I have qutebrowser v1.4.1 installed via `tox -e mkvenv-pypi` on Ubuntu Bionic (full version info below). I have all the gstreamer plugins installed and I can watch videos on youtube. However, videos with MP4 extensions are shared with me in OneDrive, and it shows an error saying "This video can't be played in this browser."

I can download the video and play it locally with mpv or vlc - no problem. vlc says the video codec is "H264 - MPEG-4 AVC (part 10)(avc1)". As far as I can see, this codec is supported by the "libav" part of the gstreamer installation. Shouldn't qutebrowser be able to play the video then?

```
qutebrowser v1.4.1
Git commit: 1d9ee2c11 (2018-07-18 09:58:49 +0200)
Backend: QtWebEngine (Chromium 65.0.3325.230)

CPython: 3.6.5
Qt: 5.11.1
PyQt: 5.11.2

sip: 4.19.12
colorama: 0.3.9
pypeg2: 2.15
jinja2: 2.10
pygments: 2.2.0
yaml: 3.13
cssutils: 1.0.2 $Id$
attr: 18.1.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: no
pdf.js: no
sqlite: 3.23.1
QtNetwork SSL: no

Style: QFusionStyle
Platform: Linux-4.15.0-24-generic-x86_64-with-Ubuntu-18.04-bionic, 64bit
Linux distribution: Ubuntu 18.04 LTS (ubuntu)
Frozen: False
Imported from /home/sturm/git/qutebrowser/qutebrowser
Using Python from /home/sturm/git/qutebrowser/.venv/bin/python3
Qt library executable path: /home/sturm/git/qutebrowser/.venv/lib/python3.6/site-packages/PyQt5/Qt/libexec, data path: /home/sturm/git/qutebrowser/.venv/lib/python3.6/site-packages/PyQt5/Qt

Paths:
cache: /home/sturm/.cache/qutebrowser
config: /home/sturm/.config/qutebrowser
data: /home/sturm/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
```

**Comment by The-Compiler** [comment]
QtWebEngine does not use GStreamer, but comes with its own codecs (from ffmpeg, via Chromium) - if you install via `tox -e mkvenv-pypi`, you get a precompiled Qt which comes without proprietary codec support.

**Comment by oliversturm** [comment]
I see. I apologize for the hassle - I read your response in stackoverflow where you were pointing out the requirement of the gstreamer plugins, but I wasn't aware that this depended on the engine.

Now, I also read the sentence in the installation instructions that mirrors what you just said. However, there isn't a hint anywhere that tells me how to use current versions of everything with a Qt version that *does* include proprietary codec support. I previously tried using `mkvenv` with no `pypi` and that leaves me with a Chromium version from the last millennium, which many sites complained about as outdated. I guess it could try to build my own Qt and then tell `tox -e mkvenv` to use that instead of the system-wide one? Or can I tell `tox -e mkvenv-pypi` to build a custom Qt on the way? Any suggestions?

**Comment by The-Compiler** [comment]
There are pretty much three options:

- Build your own Qt and PyQt from source (which will take a lot of time, disk space, and at least 8GB of RAM, maybe even 16GB)
- Use a distribution which keeps your Qt up-to-date (though at least Ubuntu Bionic seems to track latest 5.9 LTS so far, which means you get backported security fixes on top of the old Chromium)
- Use the Flatpak @glatzor built, see https://github.com/glatzor/flathub/tree/org.qutebrowser.qutebrowser (I wanted to try it out for a while, but still didn't get to it :blush:)

**Comment by oliversturm** [comment]
> you get backported security fixes on top of the old Chromium

I assume this is true, but it doesn't seem to reflect in the Chromium version. The version was showing 56.x using Bionic Qt, and I encountered websites on a daily basis that warned me about my outdated Chrome or even denied access for that reason.

I don't mind building my own Qt and PyQt. Assuming I would install this in /usr/local, will it be picked up by `tox -e mkvenv` automatically?

**Comment by The-Compiler** [comment]
> but it doesn't seem to reflect in the Chromium version

Yeah - because it *is* Chromium 56 with some additional patches on top - which means, unfortunately, you'll probably indeed miss out on *some* security fixes as Qt can't possibly track all security-relevant changes.

> I don't mind building my own Qt and PyQt. Assuming I would install this in /usr/local, will it be picked up by `tox -e mkvenv` automatically?

Only if it's picked up by your system Python (which `tox -e mkvenv` uses). Alternatively, you can do this (roughly):

- [Create](https://packaging.python.org/tutorials/installing-packages/#creating-virtual-environments) a virtualenv
- Build Qt (make sure you [turn on](https://doc.qt.io/qt-5/qtwebengine-features.html#audio-and-video-codecs) proprietary codec support)
- Build and install [sip](https://www.riverbankcomputing.com/software/sip/download) with something like `~/path/to/venv/bin/python configure.py --sip-module=PyQt5.sip && make -j4 && sudo make install`
- Build and install PyQt5 with something like `~/path/to/venv/bin/python3 configure.py --qmake ~/path/to/qt/qtbase/bin/qmake --sip ~/path/to/venv/bin/sip --confirm-license --verbose  && make -j4 && sudo make install`
- Install qutebrowser via `~/path/to/venv/bin/pip install -e .` inside the qutebrowser repo
- `~/path/to/venv/bin/qutebrowser`

**Comment by mschilli87** [comment]
> * Build your own Qt and PyQt from source (which will take a lot of time, disk space, and at least 8GB of RAM, maybe even 16GB)

As a Gentoo user I do that more or less regularly (:wink:) and my machine has 'only' 8 GB of RAM. Building everything for qutebrowser definitely is an overnight job though.
Also, I don't know to what extend I *don't* build certain features of (Py)Qt based on the use-flags. So 'blindy' building *everything* could take more than 8 GB.

**Comment by The-Compiler** [comment]
All I know is that a developer build with debug symbols takes around 12 GB of free RAM, and that 3 GB (maximum for 32-bit) isn't enough.

But good to hear that 8 GB is enough for a regular build then (oh the times we live in... :laughing:)!

**Comment by oliversturm** [comment]
Hehe, times indeed... my dev machine is used to run virtual machines in parallel and fun stuff like that, and has 128 GB of RAM. I'll give all this a try and let you know how it goes.

**Comment by oliversturm** [comment]
I have built a complete Qt, which took ~85 minutes on my machine. Built and installed sip, no problem. Building PyQt5 is giving me grief right now: https://github.com/baoboa/pyqt5/issues/44 - not something you've come across before, by any chance?

**Comment by The-Compiler** [comment]
I'm jealous! I've never seen that error, but note that that repository is in no way official - you might have better luck posting to their [mailinglist](https://www.riverbankcomputing.com/mailman/listinfo/pyqt). I'd also recommend adding the compiler/version you're using.

**Comment by oliversturm** [comment]
A mailing list? How quaint :) Thanks for the hint. Somehow I expect everybody to use github these days, or at least something very much like it.

**Comment by The-Compiler** [comment]
I assume SourceForge but with no public repo or bugtracker does not count? :laughing: PyQt exists since 1998 or so, and while they probably did a few VCS switches (CVS -> SVN -> hg or so?), the whole development workflow is quite arcane and old unfortunately...

**Comment by oliversturm** [comment]
SCCS -> CVS -> SVN -> hg was exactly what I did... not counting VSS, mercifully for a short time only and never my choice. But then I went to git since most others seemed to do that, too. I actually imported a bunch of repos from hg, which went very well. 
Anyway, it's not mainly about technology, is it? Open Source is just so much more fun if getting involved is easy...

**Comment by jgkamat** [comment]
Oli Sturm writes:

> I have built a complete Qt, which took ~85 minutes on my machine. Built and
> installed sip, no problem. Building PyQt5 is giving me grief right now:
> baoboa/pyqt5#44 - not something you've come across before, by any chance?

FWIW, I'd recommend using the distro packages provided by ubuntu bionic,
building everything yourself is a real pain unless you are using Gentoo or
some other automated system.

https://packages.ubuntu.com/search?keywords=python3-pyqt5.qtwebengine

In some ways, building yourself is worse security wise, since it's up to you
(rather than a maintainer) to ensure it is built correctly and kept it up to
date.

I don't think 80 minutes is typical for a qt build (for me at least), I would
double check whether you are building the webengine module, which (for me) is
most of the compile time.

> I assume this is true, but it doesn't seem to reflect in the Chromium
> version. The version was showing 56.x using Bionic Qt, and I encountered
> websites on a daily basis that warned me about my outdated Chrome or even
> denied access for that reason.

These can be usually bypassed by spoofing your user agent with
content.headers.user_agent. Last time I checked slack still worked flawlessly
with chrome 49 with spoofing, despite hard blocking those users. Some
applications (like riot) legitimately use new features though, and might
break.

**Comment by oliversturm** [comment]
> I'd recommend using the distro packages provided by ubuntu bionic, building everything yourself is a real pain

I'll have to read up on that. I believe I can use newer sources with an existing package (uupdate?), and there's also a way of using different configure options. But how do I make sure that the new package version installs in parallel to the standard one and packages that depend on the standard version don't require a rebuild and reliably keep using the standard version? I don't know the answers to these questions right now.

> I would double check whether you are building the webengine module

Yes, I went through configure options quite carefully before I built. webengine did take most of the time, I saw it building its ~16800 files or so.

> bypassed by spoofing your user agent with content.headers.user_agent

Well, this assumes that these sites that complain don't have any legitimate reasons to do so. I think Chromium version 56 (that's what the standard Ubuntu package reports) was released towards the end of 2016. I'd rather stay more current than that, and while they may be selectively back-porting relevant patches, there isn't any obvious way for me (or a website) to confirm this.
And of course, if I fake the version number, sites that legitimately require a certain version would be fooled into believing they can work correctly, and then fail in unique and creative ways that are completely up to me to debug...

>  it's up to you (rather than a maintainer) to ensure it is built correctly and kept it up to date

That is totally true and I'm fully aware what a pain that can be. Right now I just want to get to the point where I can see that/whether everything works perfectly if I follow this path. I would then look into some kind of automation.

**Comment by The-Compiler** [comment]
> SCCS -> CVS -> SVN -> hg was exactly what I did... not counting VSS, mercifully for a short time only and never my choice. But then I went to git since most others seemed to do that, too. I actually imported a bunch of repos from hg, which went very well.

I try to avoid hg (I'm just too used to git), and converted the company I used to work at from SVN (with an ancient server which didn't support up-to-date SVN clients) to git + GitLab.

> Anyway, it's not mainly about technology, is it? Open Source is just so much more fun if getting involved is easy...

Yeah - but technology definitely plays a big role in that. Probably the main reason qutebrowser is on GitHub rather than [my own cgit server](https://g.cmpl.cc/) where I host random other projects.

> I'll have to read up on that. I believe I can use newer sources with an existing package (uupdate?), and there's also a way of using different configure options. But how do I make sure that the new package version installs in parallel to the standard one and packages that depend on the standard version don't require a rebuild and reliably keep using the standard version? I don't know the answers to these questions right now.

There's probably no good way to do this. Probably easier/better to just use the Flatpak at that point.

> I don't think 80 minutes is typical for a qt build (for me at least), I would
double check whether you are building the webengine module, which (for me) is
most of the compile time.

Judging from "my dev machine is used to run virtual machines in parallel and fun stuff like that, and has 128 GB of RAM" I assume @oliversturm also has a similarly powerful CPU, where this definitely can be possible. A full build (with debug symbols) takes 30-45min on the server I can use in the local hackerspace (80 cores, 320 GB RAM :wink:). Can't easily use that for debug builds I need locally though... I probably should look into setting up distcc one day :laughing:

**Comment by oliversturm** [comment]
> a similarly powerful CPU

It's a Core i7-6800K @ 3.4GHz in this machine. 6/12 cores, I was running `make -j 8`. 

![image](https://user-images.githubusercontent.com/4730969/42996254-e984e652-8c0a-11e8-9981-efde88336242.png)

> Probably easier/better to just use the Flatpak at that point.

I just looked at this option seriously. I haven't used flatpak before - was able to install flatpak itself easily enough on my system. But now what? The instructions you link to suggest building the image myself - is that the way of choice? I went to flathub.org and was surprised to find that there was no pre-built qutebrowser image there. Or am I missing something?

Of course using Flatpak might still be a good way of keeping everything separate from the rest of the system. But I thought it would also save on build/rebuild time in the long term.

**Comment by The-Compiler** [comment]
Hmm, I thought that @glatzor wanted to upload it to Flathub at some point. Once it is, we certainly should add it as an alternative to the install instructions!

**Comment by oliversturm** [comment]
Well it turns out the flatpak is also not very current. At the start of the build, it downloads PyQt 5.10 and qutebrowser 1.3.0...

**Comment by glatzor** [comment]
I plan to update the flatpak soon - busy real life.

**Comment by jgkamat** [comment]
Oli Sturm writes:

> Okay, I'm out of ideas right now. I tried to have a look at the build
> configuration in the Ubuntu source package for PyQt, but there is no
> straight-forward way for me to tell how their build differs from the
> standard one (which doesn't work due to compiler errors). I guess more
> investigation is needed to find out why the build breaks on Ubuntu Bionic -
> until then, I guess I'm stuck.

This link has the ubuntu pyqt5 file along with the patches applied in a
section on the right:

https://packages.ubuntu.com/bionic/pyqt5-dev

This is the bug for one of the patches:

https://bugs.launchpad.net/ubuntu/+source/pyqt5/+bug/1766092

Since you can run an older version of PyQt5 with a newer Qt, you might be able
to just recompile ubuntu's version against your qt install...

qtwebengine 5.11 along with pyqt5 5.11 are in debian experimental right now,
so you could take a look at the sources from there as well:

https://packages.debian.org/experimental/qtwebengine5-dev

> and while they may be selectively back-porting relevant patches, there isn't
> any obvious way for me (or a website) to confirm this.

You can! Just check to see if your source release matches the one provided
from qt, and ensure the version from qt has the security patches that you
personally care about:

http://code.qt.io/cgit/qt/qtwebengine.git/log/?qt=grep&q=security&h=5.9.6

If you couldn't audit qtwebengine on a free distro, that would be a bit
worrying :)

**Comment by The-Compiler** [comment]
You can also check http://code.qt.io/cgit/qt/qtwebengine-chromium.git/log/?h=56-based to see the actual backported commits FWIW.

**Comment by mschilli87** [comment]
I just rebuilt qtwebengine on my 6-year old Gentoo notebook and it took almost exactly 4.5 h at max. usage (`make -n 5` on a dual-core i5-2520M @ 2.50GHz) peaking at ~ 5.5 GB of RAM usage (system incl. `/tmp` on tmpfs + build process).
PyQt took 'only' 30 min to build (for both, Python 2.7, and 3.5) and never consumed even 2 GB of RAM (again, incl. running my system) in the process. :smile:

**Comment by oliversturm** [comment]
Alright.... success with the build. Turns out the problem was that my Qt build was using `-qtnamespace` - I added this to the configure run because at some stage I received a warning from configure saying that I needed to do so. I eventually figured out that this broke the PyQt build. When I tried my full configure command without the option, I didn't see a warning anymore. Duh. With Qt rebuilt with the default namespace, PyQt built fine and I was able to set up qutebrowser based on the instructions above. At a glance, playing videos works now where it didn't before - I'll see if this holds up.
